### PR TITLE
feat(material/form-field-custom-control): custom phone input autoadvance

### LINKED ISSUE #13195: Custom Form Field (phone example bugs/feature enhancements)
#### Bug/Feature enhancement
In this provided example of a phone number: [creating-a-custom-form-field-control](https://material.angular.io/guide/creating-a-custom-form-field-control)
There is a few things that are lacking: 

- Size is not enforced.  the Demo has size 3 but a user is allowed to type as many char as they want.
-- The extra char's (more than 3) are overflow hidden so you dont see them. 
-- After writing this up, I found this should be `maxlength="3"`
- Auto advance. It would be nice if the input would auto advance to the next into once the size has been reached. 
- The span dashes are selectable (implementation css could turn this off).
- Since this is actually one form control, it would be nice to highlight all number at once and hit delete to clear out the values (but dont know if this is possible since they are separate input boxes)
- The demo works in the link but when you click open in stackblitz ... the HTML file is empty ... therefore the demo does not work

### PR DESCRIPTION
**Pull request by Wildhammer**
Automatically focus the next input when done with the previous one. Allow autofocusing backward when pressing backspace. onContainerClick now focuses on the last filled input on click. Using maxLength on the inputs to not allow more characters than the size of the input.
https://medium.com/@wildhammer/angular-material-phone-input-2283c7bc3372

Fixes #13195

### CODE REVIEW

File: src/components-examples/material/form-field/form-field-custom-control/form-field-custom-control-example.ts [hunk 1]
```diff
@@ -115,6 +164,21 @@ export class MyTelInput implements ControlValueAccessor, MatFormFieldControl<MyT
     }
   }
 
+  autoFocusNext(
```
Comment (mmalerba) [inline src/components-examples/material/form-field/form-field-custom-control/form-field-custom-control-example.ts:167]: nit: the name, args, return type can be all one line

File: src/components-examples/material/form-field/form-field-custom-control/form-field-custom-control-example.ts [hunk 2]
```diff
@@ -115,6 +164,21 @@ export class MyTelInput implements ControlValueAccessor, MatFormFieldControl<MyT
...
+    control: AbstractControl,
+    nextElement: HTMLInputElement | null
```
Comment (mmalerba) [inline src/components-examples/material/form-field/form-field-custom-control/form-field-custom-control-example.ts:169]: why not `nextElement?: HTMLInputElement` instead

File: src/components-examples/material/form-field/form-field-custom-control/form-field-custom-control-example.ts [hunk 3]
```diff
@@ -146,7 +216,8 @@ export class MyTelInput implements ControlValueAccessor, MatFormFieldControl<MyT
     this.disabled = isDisabled;
   }
 
-  _handleInput(): void {
+  _handleInput(control: AbstractControl, nextElement: HTMLInputElement): void {
```
Comment (mmalerba) [inline src/components-examples/material/form-field/form-field-custom-control/form-field-custom-control-example.ts:219]: make `nextElement` optional

File: src/components-examples/material/form-field/form-field-custom-control/example-tel-input-example.html [hunk 4]
```diff
@@ -1,7 +1,8 @@
...
+  <input class="example-tel-input-element" formControlName="area" size="3" 
+  maxLength="3" aria-label="Area code" (input)="_handleInput(parts.controls.area,exchange)"  #area>
   <span class="example-tel-input-spacer">&ndash;</span>
-  <input class="example-tel-input-element" formControlName="exchange" size="3" aria-label="Exchange code" (input)="_handleInput()">
+  <input class="example-tel-input-element" formControlName="exchange" maxLength="3" size="3" aria-label="Exchange code" (input)="_handleInput(parts.controls.exchange,subscriber)" (keyup.backspace)="autoFocusPrev(parts.controls.exchange,area)" #exchange>
   <span class="example-tel-input-spacer">&ndash;</span>
-  <input class="example-tel-input-element" formControlName="subscriber" size="4" aria-label="Subscriber number" (input)="_handleInput()">
+  <input class="example-tel-input-element" formControlName="subscriber" maxLength="4" size="4" aria-label="Subscriber number" (input)="_handleInput(parts.controls.subscriber,subscriber)" (keyup.backspace)="autoFocusPrev(parts.controls.subscriber,exchange)" #subscriber>
```
Comment (mmalerba) [inline src/components-examples/material/form-field/form-field-custom-control/example-tel-input-example.html:7]: `_handleInput` doesn't need a second argument in this case

File: src/components-examples/material/form-field/form-field-custom-control/example-tel-input-example.html [hunk 5]
```diff
@@ -1,7 +1,8 @@
 <div [formGroup]="parts" class="example-tel-input-container">
-  <input class="example-tel-input-element" formControlName="area" size="3" aria-label="Area code" (input)="_handleInput()">
+  <input class="example-tel-input-element" formControlName="area" size="3" 
+  maxLength="3" aria-label="Area code" (input)="_handleInput(parts.controls.area,exchange)"  #area>
```
[the remark this answers is not in the document]
Comment (Wildhammer) [reply src/components-examples/material/form-field/form-field-custom-control/example-tel-input-example.html]: resolved

File: src/components-examples/material/form-field/form-field-custom-control/form-field-custom-control-example.ts (hunk 1 above)
Comment (Wildhammer) [reply src/components-examples/material/form-field/form-field-custom-control/form-field-custom-control-example.ts]: resolved

File: src/components-examples/material/form-field/form-field-custom-control/form-field-custom-control-example.ts (hunk 2 above)
Comment (Wildhammer) [reply src/components-examples/material/form-field/form-field-custom-control/form-field-custom-control-example.ts]: resolved

File: src/components-examples/material/form-field/form-field-custom-control/form-field-custom-control-example.ts (hunk 3 above)
Comment (Wildhammer) [reply src/components-examples/material/form-field/form-field-custom-control/form-field-custom-control-example.ts]: resolved

File: src/components-examples/material/form-field/form-field-custom-control/example-tel-input-example.html (hunk 4 above)
Comment (Wildhammer) [reply src/components-examples/material/form-field/form-field-custom-control/example-tel-input-example.html]: resolved

**Comment by Wildhammer** [comment]
@mmalerba thanks for your feedback. I applied the requested changes.

File: src/components-examples/material/form-field/form-field-custom-control/form-field-custom-control-example.ts [hunk 6]
```diff
@@ -115,6 +164,18 @@ export class MyTelInput implements ControlValueAccessor, MatFormFieldControl<MyT
     }
   }
 
+  autoFocusNext(control: AbstractControl, nextElement?: HTMLInputElement): void {
+    if (!control.errors && !!nextElement) {
```
Comment (mmalerba) [inline src/components-examples/material/form-field/form-field-custom-control/form-field-custom-control-example.ts:168]: nit: `!!` is not necessary

Comment (Wildhammer) [reply src/components-examples/material/form-field/form-field-custom-control/form-field-custom-control-example.ts]: @mmalerba done!

**Pull request merged**